lib/instrumentation/index.js

```diff
@@ -70,8 +68,20 @@ Instrumentation.prototype.start = function () {
 
 Instrumentation.prototype.addEndedTransaction = function (transaction) {
   if (this._started) {
+    var queue = this._queue
     debug('adding transaction to queue %o', {id: transaction.id})
-    this._queue.add(transaction)
+    // encode the transaction as early as possible in an attempt to free up
+    // objects for garbage collection
+    queue.add(new Promise(function (resolve, reject) {
```

Comment:
We haven't used promises anywhere else in the agent. But I had some race conditions in the tests when the transaction was added to the queue asynchronously when calling `addEndedTransaction`. I felt it was easier to reason about if the queue is bumped synchronously instead of waiting till the `transaction._encode` callback is called.